README.md
# xalloc

[<img src="https://docs.rs/xalloc/badge.svg" alt="docs.rs">](https://docs.rs/xalloc/)

Dynamic suballocators for external memory (e.g., Vulkan device memory).

## Provided Algorithms

### Generic

|               Name              | Time Complexity |  Space Complexity |
| ------------------------------- | --------------- | ----------------- |
| TLSF (Two-Level Segregated Fit) | `O(1)`          | `O(N + log size)` |
| Free space bitmap               | `O(size)`       | `O(size)`         |

### Specialized

|               Name              | Time Complexity |  Space Complexity |
| ------------------------------- | --------------- | ----------------- |
| Ring buffer                     | `O(1)`          | `O(N)`            |

(`size`: heap size measured by the number of allocation units, `N`: number of allocations)

## Examples

```rust
use xalloc::{SysTlsf, SysTlsfRegion};
let mut tlsf = xalloc::SysTlsf::new(8u32);

// Allocate regions
let alloc1: (SysTlsfRegion, u32) = tlsf.alloc(4).unwrap();
let alloc2: (SysTlsfRegion, u32) = tlsf.alloc(4).unwrap();
let (region1, offset1) = alloc1;
let (region2, offset2) = alloc2;
println!("allocated #1: {:?}", (&region1, offset1));
println!("allocated #2: {:?}", (&region2, offset2));

// Deallocate a region
tlsf.dealloc(region1).unwrap();

// Now we can allocate again
tlsf.alloc(2).unwrap();
tlsf.alloc(2).unwrap();
```

## Feature Flags

- `nightly` — Enables optimizations which currently require a Nightly Rust
  compiler. This flag is now unused due to the [stabilization] of `NonNull`
  in Rust 1.25.

[stabilization]: https://blog.rust-lang.org/2018/03/29/Rust-1.25.html


License: MIT


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.7] - 2020-12-09

- Get rid of illegal uses of `std::mem::uninitialized`, replacing them with `MaybeUninit`

## [0.2.6] - 2019-02-28

- Add `Ring::dealloc_front_until` and `Ring::dealloc_back_until`

## [0.2.5] - 2018-12-05

- Add `Ring`, a circular memory allocator
- Wrap `allow(clippy::*)` with `cfg_attr(feature = "cargo-clippy", ...)` in order to make it compatible with the current stable toolchain

## [0.2.4] - 2018-11-07

- Fix a compilation issue that happens when multiple versions of `num-*` crates are mixed up
- Optimize for performance through a use of the `unreachable` crate and an architecture-specific micro-tuning

## [0.2.3] - 2018-04-02

- Use `std::ptr::NonNull` stabilized in [Rust 1.25]

[Rust 1.25]: https://blog.rust-lang.org/2018/03/29/Rust-1.25.html

## [0.2.2] - 2017-10-28

- Make more types `Debug`
- Add type aliases `SysTlsfRegion` and `SafeTlsfRegion`

## [0.2.1] - 2017-10-28

- Fix a link in `README.md`

## [0.2.0] - 2017-10-28

- Rename `TlsfSuballocRegion` to `TlsfRegion`.
- Add a free space bitmap-based allocator.
- Add CHANGELOG to track notable changes among versions.

## 0.1.0 - 2017-10-28

- Initial release.

[Unreleased]: https://github.com/yvt/xalloc-rs/compare/HEAD...v0.2.7
[0.2.7]: https://github.com/yvt/xalloc-rs/compare/v0.2.7...v0.2.6
[0.2.6]: https://github.com/yvt/xalloc-rs/compare/v0.2.6...v0.2.5
[0.2.5]: https://github.com/yvt/xalloc-rs/compare/v0.2.5...v0.2.4
[0.2.4]: https://github.com/yvt/xalloc-rs/compare/v0.2.4...v0.2.3
[0.2.3]: https://github.com/yvt/xalloc-rs/compare/v0.2.3...v0.2.2
[0.2.2]: https://github.com/yvt/xalloc-rs/compare/v0.2.2...v0.2.1
[0.2.1]: https://github.com/yvt/xalloc-rs/compare/v0.2.1...v0.2.0
[0.2.0]: https://github.com/yvt/xalloc-rs/compare/v0.2.0...v0.1.0


src/lib.rs
//
// Copyright 2017 yvt, all rights reserved.
//
// Licensed under the MIT license <LICENSE-MIT or
// http://opensource.org/licenses/MIT>. This file may
// not be copied, modified,or distributed except
// according to those terms.
//
//! Dynamic suballocators for external memory (e.g., Vulkan device memory).
//!
//! # Provided Algorithms
//!
//! ## Generic
//!
//! |               Name              | Time Complexity |  Space Complexity |
//! | ------------------------------- | --------------- | ----------------- |
//! | TLSF (Two-Level Segregated Fit) | `O(1)`          | `O(N + log size)` |
//! | Free space bitmap               | `O(size)`       | `O(size)`         |
//!
//! ## Specialized
//!
//! |               Name              | Time Complexity |  Space Complexity |
//! | ------------------------------- | --------------- | ----------------- |
//! | Ring buffer                     | `O(1)`          | `O(N)`            |
//!
//! (`size`: heap size measured by the number of allocation units, `N`: number of allocations)
//!
//! # Examples
//!
//! ```
//! use xalloc::{SysTlsf, SysTlsfRegion};
//! let mut tlsf = xalloc::SysTlsf::new(8u32);
//!
//! // Allocate regions
//! let alloc1: (SysTlsfRegion, u32) = tlsf.alloc(4).unwrap();
//! let alloc2: (SysTlsfRegion, u32) = tlsf.alloc(4).unwrap();
//! let (region1, offset1) = alloc1;
//! let (region2, offset2) = alloc2;
//! println!("allocated #1: {:?}", (&region1, offset1));
//! println!("allocated #2: {:?}", (&region2, offset2));
//!
//! // Deallocate a region
//! tlsf.dealloc(region1).unwrap();
//!
//! // Now we can allocate again
//! tlsf.alloc(2).unwrap();
//! tlsf.alloc(2).unwrap();
//! ```
//!
//! # Feature Flags
//!
//! - `nightly` — Enables optimizations which currently require a Nightly Rust
//!   compiler. This flag is now unused due to the [stabilization] of `NonNull`
//!   in Rust 1.25.
//!
//! [stabilization]: https://blog.rust-lang.org/2018/03/29/Rust-1.25.html
//!
// Clippy does not understand that generic numeric types are not always
// as capable as built-in ones and raise false warnings
#![cfg_attr(feature = "cargo-clippy", allow(clippy::op_ref))]

pub extern crate num;
extern crate unreachable;

pub mod arena;
pub mod bitmap;
mod bitmaputils;
pub mod int;
pub mod ring;
pub mod tlsf;

pub use self::bitmap::{BitmapAlloc, BitmapAllocRegion};
pub use self::ring::{Ring, RingRegion};
pub use self::tlsf::{
    SafeTlsf, SafeTlsfRegion, SysTlsf, SysTlsfRegion, Tlsf, TlsfBlock, TlsfRegion,
};
